src/shogun/io/SGIO.h

```diff
@@ -493,40 +415,16 @@ class SGIO
 		 */
 		static uint32_t ss_length(substring s);
 
-		/** increase reference counter
-		 *
-		 * @return reference count
-		 */
-		int32_t ref();
-
-		/** display reference counter
-		 *
-		 * @return reference count
-		 */
-		int32_t ref_count() const;
-
-		/** decrement reference counter and deallocate object if refcount is zero
-		 * before or after decrementing it
-		 *
-		 * @return reference count
-		 */
-		int32_t unref();
-
 		/** @return object name */
 		inline const char* get_name() { return "SGIO"; }
 
-	protected:
-		/** get message intro
-		 *
-		 * @param prio message priority
-		 * @return message intro or NULL if message is not to be
-		 *         printed
-		 */
-		const char* get_msg_intro(EMessageType prio) const;
+	private:
+		/** Prints a formatted message */
+		void message_(EMessageType prio, const SourceLocation& loc, const fmt::string_view& msg) const;
+
+		/** Updates log pattern */
+		void update_pattern();
```

Comment:
pattern is `[date time {source_location} loglevel] msg`
example:
`[07/27/19 23:39:33 debug] Comparing parameter DynamicObjectArray::dim1_size of type int.`